automation/taskcluster/rename_snapshot_artifacts.py

```diff
@@ -0,0 +1,126 @@
+
+
+SNAPSHOT_TIMESTAMP_REGEX = r'\d{8}\.\d{6}-\d{1}'
+
+
+def _extract_and_check_timestamps(archive_filename, regex):
+    """Function to check if a timestamp-based regex matches a given filename"""
+    match = re.search(regex, archive_filename)
+    try:
+        identifier = match.group()
+    except AttributeError:
+        raise Exception(
+            'File "{}" present in archive has invalid identifier. '
+            'Expected YYYYMMDD.HHMMSS-BUILDNUMBER within in'.format(
+                archive_filename
+            )
+        )
+    timestamp, buildno = identifier.split('-')
+    try:
+        datetime.datetime.strptime(timestamp, '%Y%m%d.%H%M%S')
+    except ValueError:
+        raise Exception(
+            'File "{}" present in archive has invalid timestamp. '
+            'Expected YYYYMMDD.HHMMSS within in'.format(
+                archive_filename
+            )
+        )
+
+    date, clock = timestamp.split('.')
+    return date, clock, buildno
+
+
+def _extract_old_timestamp(files):
+    """Function to find and ensure that a single timestamp has been used
+    across a component's related files"""
+    identifiers_collection = set()
+    for file_ in files:
+        (date, clock, bno) = _extract_and_check_timestamps(file_,
+                                                           SNAPSHOT_TIMESTAMP_REGEX)
+        _args = {
+            'date': date,
+            'clock': clock,
+            'buildno': bno
+        }
+        identifiers_collection.add(frozenset(_args.items()))
+
+    if len(identifiers_collection) > 1:
+        # bail if there are different timestamps/buildnumbers across the files
+        raise Exception('Different buildnumbers/timestamps identified within the files')
+    elif len(identifiers_collection) == 0:
+        # bail if no valid timestamp was found
+        raise Exception('No valid timestamp was identified within the files')
+
+    old_timestamp_dict = dict(next(iter(identifiers_collection)))
+    return '{date}.{clock}-{buildno}'.format(**old_timestamp_dict)
+
+
+def is_snapshot_timestamped_file(filename):
+    """Function to filter out filenames that are part of the snapshot releae but
+    they are not timestamped."""
+
+    for extension, hash_extension in itertools.product(AAR_EXTENSIONS, HASH_EXTENSIONS):
+        if filename.endswith(extension + hash_extension):
+            return True
+    return False
+
+
+def process_snapshots_artifacts(path, timestamp):
+    files = []
+    # gather all the files recursively
+    for (dirpath, dirnames, filenames) in os.walk(args.path):
+        if filenames:
+            files.extend(filenames)
+
+    # exclude files that are not timestamped such as metadata.xml
+    files = filter(lambda f: is_snapshot_timestamped_file(f), files)
+
+    # extract their timestamps
+    old_timestamp = _extract_old_timestamp(files)
+
+    # walk recursively again and rename the files accordingly
+    for (dirpath, dirnames, filenames) in os.walk(args.path):
```

Comment:
Nit: `for (dirpath, _, filenames) in os.walk(path):`

Reply:
👍